odd-platform-api/src/main/java/org/opendatadiscovery/oddplatform/service/DatasetStructureServiceImpl.java

```diff
@@ -0,0 +1,153 @@
+                }
+                return reactiveDatasetVersionRepository.getPenultimateVersions(latestVersionsWithPenultimates)
+                    .defaultIfEmpty(List.of())
+                    .flatMap(penultimateList -> getLastStructureDelta(
+                        latestVersionsWithPenultimates, penultimateList))
+                    .map(map -> Tuples.of(map, dataStructure));
+            });
+    }
+
+    @Override
+    public Mono<Map<String, DatasetStructureDelta>> getLastStructureDelta(final List<DatasetVersionPojo> versions,
+                                                                          final Set<Long> dataVersionPojoIds) {
+        return reactiveDatasetVersionRepository.getDatasetVersionPojoIds(dataVersionPojoIds)
+            .flatMap(vidToFields -> {
+                final Map<String, List<DatasetVersionPojo>> dsOddrnToVersions = versions
+                    .stream()
+                    .collect(Collectors.groupingBy(DatasetVersionPojo::getDatasetOddrn));
+
+                return Mono.just(dsOddrnToVersions.entrySet().stream()
+                    .map(e -> {
+                        final List<DatasetVersionPojo> v = e.getValue().stream()
+                            .sorted(Comparator.comparing(DatasetVersionPojo::getVersion))
+                            .collect(Collectors.toList());
+
+                        return Pair.of(e.getKey(), new DatasetStructureDelta(
+                            vidToFields.get(v.get(0).getId()),
+                            vidToFields.get(v.get(1).getId())
+                        ));
+                    })
+                    .collect(Collectors.toMap(Pair::getLeft, Pair::getRight)));
+            });
+    }
+
+    @NotNull
+    private Mono<Map<String, DatasetStructureDelta>> getLastStructureDelta(
+        final List<DatasetVersionPojo> latestVersions,
+        final List<DatasetVersionPojo> penultimateList) {
+        final List<DatasetVersionPojo> versions = ListUtils.union(latestVersions, penultimateList);
+        final Set<Long> dataVersionPojoIds = versions.stream()
+            .map(DatasetVersionPojo::getId)
+            .collect(Collectors.toSet());
+        return getLastStructureDelta(versions, dataVersionPojoIds);
+    }
+
+    @Override
+    public Mono<List<DatasetVersionPojo>> getVersions(final Map<String, EnrichedDataEntityIngestionDto> datasetDict,
+                                                      final Set<Long> datasetIds) {
+        return reactiveDatasetVersionRepository
+            .getLatestVersions(datasetIds)
+            .map(fetchedVersions -> fetchedVersions.stream()
+                .map(fetchedVersion -> incrementVersion(datasetDict, fetchedVersion))
+                .filter(Objects::nonNull)
+                .collect(Collectors.toList()));
+    }
+
+    private DatasetVersionPojo incrementVersion(final Map<String, EnrichedDataEntityIngestionDto> datasetDict,
```

Comment:
I know that this is old code, but I also think that it can be improved, because method which is called incrementVersion actually doesn't just increment version, it returns null if hash is the same.
As for me, we should do this filter above, not inside method called "incrementVersion"